### PR TITLE
Add a js_class to implement the Class trait without boilerplate

### PR DESCRIPTION
**Pull request by hansl**
This also adds a JsInstance that verifies that `this` is of the proper class, and an `Ignore` that ignore arguments.

The syntax is a bit special because of limitations of macro_rules. For example, the way fields are defined. It was impossible to keep both assignments (e.g. `public field = 123;`) and dynamic fields.

This reduces significantly the boilerplate for declaring classes. The example in class.rs is more than twice as long (if you remove comments) than the macro is.

### CODE REVIEW

**Review by jedel1043 — COMMENTED** [review]
Really impressive work! Definitely having a boilerplate-free way to create JsClasses is a thing we should have.

I was wondering, what is the benefit of doing it with a macro_rules vs a procedural macro? I'm just worried about using macro_rules for this since it would pretty much disable all IDE autocompletions vs a procedural macro, which is a lot nicer to IDEs.

I was envisioning something like wasm_bindgen's [`getter` and `setter` attributes](https://rustwasm.github.io/docs/wasm-bindgen/reference/attributes/on-rust-exports/getter-and-setter.html)

**Comment by hansl** [comment]
It was just easier for me to build a macro, and I wanted to have a more JS-ey way to describe the class. It would make sense to have a proc macro at some point, I just don't have the time yet. This reduces a lot of code already.

**Comment by jedel1043** [comment]
> It was just easier for me to build a macro, and I wanted to have a more JS-ey way to describe the class. It would make sense to have a proc macro at some point, I just don't have the time yet. This reduces a lot of code already.

Fail enough. I think we can start by offering a macro rules, then switch to a proc macro in the future.

File: core/interop/src/lib.rs [hunk 1]
```diff
@@ -305,6 +325,74 @@ impl<'a, T: TryFromJs> TryFromJsArgument<'a> for JsThis<T> {
...
+}
+
+/// Captures a class instance from the `this` value in a JS function. The class
+/// will be a non-mutable reference of Rust type `T`, if it is an instance of `T`.
+///
+/// To have more flexibility on the parsing of the `this` value, you can use the
+/// [`JsThis`] capture instead.
+#[derive(Debug, Clone)]
+pub struct JsClass<T: NativeObject> {
+    inner: boa_engine::JsObject,
```
Comment (jedel1043) [inline core/interop/src/lib.rs:343]: You can use `boa_engine::JsObject<T>` and `JsObject::downcast` instead, which gives us infallible downcasts after the first downcast.

File: core/interop/src/lib.rs [hunk 2]
```diff
@@ -305,6 +325,74 @@ impl<'a, T: TryFromJs> TryFromJsArgument<'a> for JsThis<T> {
...
+    ) -> JsResult<(Self, &'a [JsValue])> {
+        if let Some(object) = this.as_object() {
+            if object.downcast_ref::<T>().is_some() {
+                return Ok((
+                    JsClass {
+                        inner: object.clone(),
+                        _ty: PhantomData,
+                    },
+                    rest,
+                ));
```
Comment (jedel1043) [inline core/interop/src/lib.rs:386]: See above. This would be a `JsObject::downcast` instead.

**Review by jedel1043 — CHANGES_REQUESTED** [review]
Thank you! I have some suggestions that remove the `unwrap`s.

File: core/interop/src/lib.rs (hunk 1 above)
Comment (hansl) [reply core/interop/src/lib.rs:343]: Sure, but it downcasts to `Object<T>`, not `T` :(

Comment (hansl) [reply core/interop/src/lib.rs:343]: I don't know how to get a `T` (or more specifically a `&mut T`) from an `Object<T>`.

Ideally I want a `GcRef<'_, T>` (and `GcRefMut<'_, _, T>`) directly.

Comment (jedel1043) [reply core/interop/src/lib.rs:343]: Nono, there's a `JsObject::downcast` that converts from `JsObject` to `JsObject<T>, which makes it possible to  access the data between borrows without having to call `downcast_mut` or `downcast_ref`

Comment (hansl) [reply core/interop/src/lib.rs:343]: Yes but that data is non-mut. Are you suggesting I should add this to `Object<T>`?

```rust
pub fn data_mut(&mut self) -> &mut T {
  self.data
}
```

Comment (jedel1043) [reply core/interop/src/lib.rs:343]: After that, if you want to map to the inner data, you can do

`GcRef::map(o.borrow(), |o| &o.data)`

Same with `GcRefMut`

Comment (jedel1043) [reply core/interop/src/lib.rs:343]: Ah right, we don't have `data_mut()`. Yeah, just add it to `Object<T>`

Comment (hansl) [reply core/interop/src/lib.rs:343]: Uh for some reason I missed `GcRef[Mut]::map`. Probably because it's still unstable in std. Thanks!

**Comment by hansl** [comment]
@jedel1043 PTAL

**Review by jedel1043 — APPROVED** [review]
Looks perfect to me!

**Review by nekevss — APPROVED** [review]
Overall, looks good to me! Thanks for all your work on this!

**Pull request merged**